fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/BundleContentsStorage.java

```diff
@@ -0,0 +1,189 @@
+		StoragePreconditions.notBlankNotNegative(resource, maxAmount);
+
+		if (!isStillBundle()) return 0;
+
+		if (maxAmount > Integer.MAX_VALUE) maxAmount = Integer.MAX_VALUE;
+
+		ItemStack stack = resource.toStack((int) maxAmount);
+
+		if (!BundleContentsComponent.canBeBundled(stack)) return 0;
+
+		var builder = new BundleContentsComponent.Builder(bundleContents());
+
+		int inserted = builder.add(stack);
+
+		ComponentChanges changes = ComponentChanges.builder()
+				.add(DataComponentTypes.BUNDLE_CONTENTS, builder.build())
+				.build();
+
+		if (!updateStack(changes, transaction)) return 0;
+
+		return inserted;
+	}
+
+	@Override
+	public long extract(ItemVariant resource, long maxAmount, TransactionContext transaction) {
+		StoragePreconditions.notNegative(maxAmount);
+
+		if (!isStillBundle()) return 0;
+
+		updateSlotsIfNeeded();
+
+		long amount = 0;
+
+		for (StorageView<ItemVariant> slot : slots) {
+			amount += slot.extract(resource, maxAmount - amount, transaction);
+			if (amount == maxAmount) break;
+		}
+
+		return amount;
+	}
+
+	@Override
+	public Iterator<StorageView<ItemVariant>> iterator() {
+		updateSlotsIfNeeded();
+
+		return slots.iterator();
+	}
+
+	private boolean isStillBundle() {
+		return ctx.getItemVariant().getItem() == Items.BUNDLE;
+	}
+
+	private void updateSlotsIfNeeded() {
+		if (slots.size() != bundleContents().size()) {
+			while (bundleContents().size() > slotCache.size()) {
+				slotCache.add(new BundleSlotWrapper(slotCache.size()));
+			}
+
+			slots = Collections.unmodifiableList(slotCache.subList(0, bundleContents().size()));
+		}
+	}
+
+	BundleContentsComponent bundleContents() {
+		return ctx.getItemVariant().getComponentMap().getOrDefault(DataComponentTypes.BUNDLE_CONTENTS, BundleContentsComponent.DEFAULT);
+	}
+
+	private class BundleSlotWrapper implements StorageView<ItemVariant> {
+		private final int index;
+
+		private BundleSlotWrapper(int index) {
+			this.index = index;
+		}
+
+		private ItemStack getStack() {
+			if (bundleContents().size() <= index) return ItemStack.EMPTY;
+
+			return ((List<ItemStack>) bundleContents().iterate()).get(index);
+		}
+
+		@Override
+		public long extract(ItemVariant resource, long maxAmount, TransactionContext transaction) {
+			StoragePreconditions.notNegative(maxAmount);
+
+			if (!BundleContentsStorage.this.isStillBundle()) return 0;
+			if (bundleContents().size() <= index) return 0;
+			if (!resource.matches(getStack())) return 0;
+
+			var stacksCopy = new ArrayList<>((Collection<ItemStack>) bundleContents().iterateCopy());
+
+			int extracted = (int) Math.min(stacksCopy.get(index).getCount(), maxAmount);
+
+			stacksCopy.get(index).decrement(extracted);
+			if (stacksCopy.get(index).isEmpty()) stacksCopy.remove(index);
+
+			var builder = new BundleContentsComponent.Builder(BundleContentsComponent.DEFAULT);
+
+			for (ItemStack stack : stacksCopy) builder.add(stack);
```

Comment:
I believe this reverses the order. `Builder#add` prepends, and the first element in `stacksCopy` is getting added first.

In fact, this shouldn't even use the builder. If you have the list of stacks, you can directly call the constructor.

Reply:
I don't remember exactly why I used the builder here. I think it's because the builder calculated the occupancy for me?

Reply:
The vanilla ContainerComponentModifier for bundles does use the builder when making a bundle component from a list of stacks, though

Reply:
Constructor should have auto-calced the occupancy. When we're removing items from the bundle, as opposed to adding it, there shouldn't be any need for the extra logic in the builder.

Reply:
Yeah, I figured that out already, and used that constructor.